Setting Code to empty in C++ panics with index out of range

Here is the stack trace:
```
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/Fantom-foundation/Carmen/go/state.(*CppState).SetCode.func1(0x8000000000009508?, 0x0?, 0xc0bef9c030?)
        /home/herbert/coding/aida/carmen/go/state/cpp_state.go:104 +0x106
github.com/Fantom-foundation/Carmen/go/state.(*CppState).SetCode(0x8ecf723d801216ac?, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
        /home/herbert/coding/aida/carmen/go/state/cpp_state.go:104 +0x57
github.com/Fantom-foundation/Carmen/go/state.(*stateDB).EndTransaction(0xc06abd4a00)
        /home/herbert/coding/aida/carmen/go/state/state_db.go:542 +0xcd3
github.com/Fantom-foundation/Aida/tracer/state.(*carmenStateDB).Finalise(0xc06aba61c0, 0x90?)
        /home/herbert/coding/aida/tracer/state/carmen.go:127 +0x29
github.com/Fantom-foundation/Aida/tracer/operation.(*Finalise).Execute(0xc0cff28aac, {0xc49c28, 0xc06aba61c0}, 0xc000233080?)
        /home/herbert/coding/aida/tracer/operation/finalise.go:44 +0x51
github.com/Fantom-foundation/Aida/tracer/operation.Execute({0xc43a88, 0xc0cff28aac}, {0xc49c28?, 0xc06aba61c0?}, 0x0?)
        /home/herbert/coding/aida/tracer/operation/operation.go:162 +0x3e
github.com/Fantom-foundation/Aida/cmd/trace-cli/trace.traceReplayTask(0x45a43a, 0x4630c0, 0x7ffc4743b5fa?)
        /home/herbert/coding/aida/cmd/trace-cli/trace/replay.go:131 +0x809
github.com/Fantom-foundation/Aida/cmd/trace-cli/trace.traceReplayAction(0xc000233400)
        /home/herbert/coding/aida/cmd/trace-cli/trace/replay.go:192 +0x297
github.com/urfave/cli/v2.(*Command).Run(0xfe8cc0, 0xc000233040)
        /home/herbert/go/pkg/mod/github.com/urfave/cli/v2@v2.19.2/command.go:177 +0x746
github.com/urfave/cli/v2.(*App).RunContext(0xc0000fe700, {0xc437b0?, 0xc0000260c8}, {0xc000024200, 0x10, 0x10})
        /home/herbert/go/pkg/mod/github.com/urfave/cli/v2@v2.19.2/app.go:389 +0x105e
github.com/urfave/cli/v2.(*App).Run(...)
        /home/herbert/go/pkg/mod/github.com/urfave/cli/v2@v2.19.2/app.go:254
main.main()
        /home/herbert/coding/aida/cmd/trace-cli/main.go:33 +0x12e
```

--- a/go/state/cpp_state.go
+++ b/go/state/cpp_state.go
@@ -92,7 +92,7 @@ func (cs *CppState) SetStorage(address common.Address, key common.Key, value com
 func (cs *CppState) GetCode(address common.Address) ([]byte, error) {
 	const max_size = 25000 // Contract limit is 24577
 	code := make([]byte, max_size)
-	var size C.uint32_t
+	var size C.uint32_t = max_size
 	C.Carmen_GetCode(cs.state, unsafe.Pointer(&address[0]), unsafe.Pointer(&code[0]), &size)
 	if size >= max_size {
 		return nil, fmt.Errorf("Unable to load contract exceeding maximum capacity of %v", max_size)
@@ -101,7 +101,11 @@ func (cs *CppState) GetCode(address common.Address) ([]byte, error) {
 }
 
 func (cs *CppState) SetCode(address common.Address, code []byte) error {
-	C.Carmen_SetCode(cs.state, unsafe.Pointer(&address[0]), unsafe.Pointer(&code[0]), C.uint32_t(len(code)))
+	var code_ptr unsafe.Pointer
+	if len(code) > 0 {
+		code_ptr = unsafe.Pointer(&code[0])
+	}
+	C.Carmen_SetCode(cs.state, unsafe.Pointer(&address[0]), code_ptr, C.uint32_t(len(code)))
 	return nil
 }
 

--- a/go/state/cpp_state_test.go
+++ b/go/state/cpp_state_test.go
@@ -1,6 +1,7 @@
 package state
 
 import (
+	"bytes"
 	"testing"
 
 	"github.com/Fantom-foundation/Carmen/go/common"
@@ -151,6 +152,74 @@ func TestWriteAndReadSlot(t *testing.T) {
 			}
 			if value != val1 {
 				t.Errorf("Invalid value read, got %v, wanted %v", value, val1)
+			}
+		})
+	}
+}
+
+func getTestCodeOfLength(size int) []byte {
+	res := make([]byte, size)
+	for i := 0; i < size; i++ {
+		res[i] = byte(i)
+	}
+	return res
+}
+
+func getTestCodes() [][]byte {
+	return [][]byte{
+		nil,
+		{},
+		{0xAC},
+		{0xAC, 0xDC},
+		getTestCodeOfLength(100),
+		getTestCodeOfLength(1000),
+		getTestCodeOfLength(24577),
+	}
+}
+
+func TestSetAndGetCode(t *testing.T) {
+	for _, named_state := range initCppStates(t) {
+		name := named_state.name
+		state := named_state.state
+		t.Run(name, func(t *testing.T) {
+			for _, code := range getTestCodes() {
+				err := state.SetCode(address1, code)
+				if err != nil {
+					t.Fatalf("Error setting code: %v", err)
+				}
+				value, err := state.GetCode(address1)
+				if err != nil {
+					t.Fatalf("Error fetching code: %v", err)
+				}
+				if !bytes.Equal(code, value) {
+					t.Errorf("Invalid value read, got %v, wanted %v", value, code)
+				}
+			}
+		})
+	}
+}
+
+func TestSetAndGetCodeHash(t *testing.T) {
+	for _, named_state := range initCppStates(t) {
+		name := named_state.name
+		state := named_state.state
+		t.Run(name, func(t *testing.T) {
+			for _, code := range getTestCodes() {
+				err := state.SetCode(address1, code)
+				if err != nil {
+					t.Fatalf("Error setting code: %v", err)
+				}
+				hash, err := state.GetCodeHash(address1)
+				if err != nil {
+					t.Fatalf("Error fetching code: %v", err)
+				}
+				want := common.GetSha256Hash(code)
+				if len(code) == 0 {
+					want = common.Hash{}
+				}
+				if hash != want {
+					t.Errorf("Invalid code hash, got %v, wanted %v", hash, want)
+				}
 			}
 		})
 	}

